File: app/src/main/java/cdreyfus/xebia_henri_potier/book/BookPresenter.java
package cdreyfus.xebia_henri_potier.book;

import android.content.Context;

import java.util.Map;

import cdreyfus.xebia_henri_potier.HenriPotierApplication;
import cdreyfus.xebia_henri_potier.basket.Basket;

public class BookPresenter {

    private View view;
    private Context context;
    private Book book;
    private Basket basket;

    BookPresenter(View view, Context context) {
        this.view = view;
        this.context = context;
    }

    public void initBook(String isbn) {
        BookDao bookDao = ((HenriPotierApplication) context).getDaoSession().getBookDao();
        this.book = bookDao.queryBuilder().where(BookDao.Properties.Isbn.eq(isbn)).build().unique();
    }

    public void initBasket(Basket mBasket) {
        this.basket = mBasket;
    }


    private void setTitre(String titre) {
        view.setTitre(titre);
    }

    private void setCover(String cover) {
        view.setCover(cover);
    }

    private void setSynopsis(String resume) {
        view.setResume(resume);
    }

    private void setPrice(float price) {
        view.setPrice(price);
    }

    public void getBook() {
        updateBook();
        updateBasketButtons();
    }

    private void updateBook() {
        setTitre(book.getTitle());
        setCover(book.getCover());
        setSynopsis(book.getSynopsis());
        setPrice(book.getPrice());
    }

    private void updateBasketButtons() {
        if (basket.getBooksQuantitiesMap().containsKey(book)) {
            view.setBookInBasket(basket.getBooksQuantitiesMap().get(book));
        } else {
            view.setNotInBasket();
        }
    }

    public void addToBasket() {
        basket.getBooksQuantitiesMap().put(book, 1);
        view.setBookInBasket(basket.getBooksQuantitiesMap().get(book));
    }

    public void removeFromBasket() {
        basket.getBooksQuantitiesMap().remove(book);
        view.setNotInBasket();
    }

    public void editQuantityinBasket(int newVal) {
        for (Map.Entry<Book, Integer> entry : basket.getBooksQuantitiesMap().entrySet()) {
            if (entry.getKey().equals(book)) {
                entry.setValue(newVal);
            }
        }
        view.setBookInBasket(newVal);
        view.hideNumberPicker();
    }

    public void selectQuantityInBasket(){
        view.showNumberPicker(book.getTitle(), basket.getBooksQuantitiesMap().get(book));
    }


    public interface View {

        void setTitre(String titre);

        void setCover(String cover);

        void setResume(String resume);

        void setPrice(float price);

        void setBookInBasket(int quantity);

        void setNotInBasket();

        void showNumberPicker(String title, int quantity);

        void hideNumberPicker();
    }

}

File: app/src/main/java/cdreyfus/xebia_henri_potier/book/BookPresenter.kt
package cdreyfus.xebia_henri_potier.book

import android.content.Intent
import cdreyfus.xebia_henri_potier.basket.Basket
import cdreyfus.xebia_henri_potier.utils.Utils

class BookPresenter internal constructor() {
    private var basket: Basket = Basket()
    private var view: View? = null
    private var book:Book = Book()

    private var quantity: Int = 0

    fun attachView(view:View){
        this.view = view
    }

    fun detachView() {
        this.view = null
    }

    fun initBook(book: Book){
        this.book = book
    }

    fun initBasket(basket: Basket) {
        this.basket = basket
        if (basket.booksQuantitiesMap.containsKey(book)) {
            quantity = basket.booksQuantitiesMap.getValue(book)
            updateBasketButtons()
        }
    }

    private fun setTitre(titre: String) {
        view?.setTitre(titre)
    }

    private fun setCover(cover: String) {
        view?.setCover(cover)
    }

    private fun setSynopsis(resume: ArrayList<String>) {
        view?.setResume(resume)
    }

    private fun setPrice(price: Float?) {
        view?.setPrice(price)
    }


    fun updateViewBook() {
        setTitre(book.title)
        setCover(book.cover)
        setSynopsis(book.synopsis)
        setPrice(book.price.toFloat())
    }

    fun updateBasketButtons() {
        if (quantity > 0) {
            view?.setBookInBasket(quantity)
        } else {
            view?.setNotInBasket()
        }
    }

    fun addToBasket() {
        quantity = 1
        basket.booksQuantitiesMap[book] = 1
        view?.setBookInBasket(quantity)

    }

    fun removeFromBasket() {
        quantity = 0
        basket.booksQuantitiesMap.remove(book)
        view?.setNotInBasket()
    }

    fun editQuantityinBasket(newVal: Int) {
        quantity = newVal
        basket.booksQuantitiesMap[book] = quantity
        view?.setBookInBasket(newVal)
        view?.hideNumberPicker()
    }

    fun selectQuantityInBasket() {
        view?.showNumberPicker(book.title, quantity)
    }

    fun sendBasketForResult(intent: Intent) {
        intent.putExtra(Utils.EXTRA_CONTENT_BASKET, basket)
    }


    interface View {

        fun setTitre(titre: String)

        fun setCover(cover: String)

        fun setResume(resume: ArrayList<String>)

        fun setPrice(price: Float?)

        fun setBookInBasket(quantity: Int)

        fun setNotInBasket()

        fun hideNumberPicker()

        fun showNumberPicker(title: String, quantity: Int)
    }

}


